AuroraClusterMonitor: Allow weighted replica's

jentfoo opened the pull request:
This adds the static function `AuroraClusterMonitor.setReplicaWeight(String, int, int)` to be able to set weights for a given server.
As described in the javadocs, the weight effectively represents how many times that server should be choosen as a random replica among the cluster.
A weight of zero will prevent the server from being used unless it's the only available healthy replica.  While larger numbers will deliever additional traffic to those instances.

In addition dependency versions were updated and JMH was added as a test dependency.  A lot of benchmarking was done to make sure that this does not impact the performance of the driver.

In order to avoid performance impacts the weights are managed in additional collection.  Making the cluster state checker do a little additional logic to manage this collection.  Allowing the server selection logic to remain almost unchanged, and with no statistically significant performance impact.

It will be easier to see the logic change in the first commit in isolation.  I decided to do some minor renames as a second commit to this PR.

This PR resolves #2

AltSysrq commented on arcCommon/src/main/java/org/threadly/db/aurora/AuroraClusterMonitor.java:
```diff
@@ -377,44 +445,128 @@ protected void checkAllServers() {
+          weightedReplicaServers = Collections.singletonList(updatedWeightedReplicaServers.get(0));
+        } else {
+          updatedWeightedReplicaServers.trimToSize();
+          weightedReplicaServers = updatedWeightedReplicaServers;
```
Given that this updates `weightedReplicaServers` all at once, wouldn't it be better to just do a `Collections.shuffle()` here and ditch the complex and inefficient random-insertion logic in `addWeightedServer()`?

jentfoo replied:
Benchmarking shows that the answer is "it depends".  If the weight is only one or two, this logic is ideal.  For higher weights you're right that a shuffle is more performant.  Part of what lead me to this is that we are guaranteed to have two well placed weights with this logic, where a shuffle is completely random.

I was thinking lower weights are probably more common, and that the overhead on this rare process is likely not a concern.  Thoughts?

AltSysrq replied:
> two well placed weights

What does this mean?

jentfoo replied:
We choose a "random" server by going through the list of replica's linearly for efficiency.  I considered when writing the weight code that having a sever repeated means that this behavior will cause spikes as it sends repeated queries to one server before it finds the next one in the list.

So by well placed I mean "as far from each other as possible"

jentfoo commented:
> In the context of #24, I'm not sure this is really what we want, mainly since it doesn't consider the master in the weighting.

How can the master be reliably considered in these weights?  If you need to write, you must go to the writer, and the availability zone is not able to be considered.

If you don't need to write, then instructing the driver to choose a random replica which will delegate to the logic in this PR.

Maybe I am not understanding how a master can be considered here?

AltSysrq commented:
> Maybe I am not understanding how a master can be considered here?

Concerete example: We have an Aurora cluster with three nodes

- foo-useast1a.rds.amazonaws.com, master
- foo-useast1d.rds.amazonaws.com, replica
- foo-useast1e.rds.amazonaws.com, replica

If a service is running in us-east-1a, we want its _reads_ to go to the RDS instance in that zone, which in this case happens to be the master.

jentfoo commented:
> If a service is running in us-east-1a, we want its reads to go to the RDS instance in that zone, which in this case happens to be the master.

Since the service already has to determine the logic for how it configures the weights.  Can't that service just make the determination to not use a replica and instead recommend the master in all cases?

AltSysrq commented:
Not easily. The master and a replica could be in the same AZ, in which case we do want to round-robin them both. The master could change at runtime, which could turn a previously optimal configuration into a pessimal configuration.

jentfoo commented:
Yeah, the application could still make a determination of how often / when it selects a master read vs a weighted replica if it wanted to consider the master in the balance.  But I still see your point about unfortunately needing to take that logic into the client.

Let me think a bit more on this, feel free to post any new ideas you have on either PR.

jentfoo commented:
@AltSysrq I pushed a commit to this PR so that you can more easily get notifications on cluster state changes.  I see one of two ways this can be utilized and want some feedback on what you're thoughts are:

1) This should provide you what you need to make the determination of if a master read should be used, or if it should delegate to a replica.

2) The second option would be to allow the `getRandomReadReplica()` logic to be replaced.  This could allow for a more generic replica server provider implementation.  I would create a new function like `setReplicaProvider(Collection<String> hosts, ReplicaProvider)` similar to how the listener is registered (with the expectation that most replica providers are also listeners).

In this case I would remove the newly added `AuroraServerClusterMonitor.setReplicaWeight` and instead have this logic implemented as a plugable ReplicaProvider (maybe the one used by default).

My take...
I don't really like the ergonomics of configuration through a `Collection<String>` representing the cluster.  But I am unsure how else to make this logic able to be cluster specific (which is the only way it makes sense).

But if we can get past that, maybe a generic solution is ideal here.  Having the library provide first class support for weighting, az relative, or other similar logic.

AltSysrq commented:
I really can't say I'm a fan of either option — either way we end up with a lot more complexity both here and in the calling code.

Taking a step back, I think the weighting approach would work for colocation if we just made the `Any` choice also work with weighting.

jentfoo commented:
> Taking a step back, I think the weighting approach would work for colocation if we just made the Any choice also work with weighting.

This is a good idea, I like it.

Do you think the cluster state listener is worth the awkwardness of how it needs to be added to the cluster?

AltSysrq commented:
> Do you think the cluster state listener is worth the awkwardness of how it needs to be added to the cluster?

Probably not right now.

jentfoo commented:
@AltSysrq Take a look when you get a chance.  I reverted my previous commit, and added a new one so that "Any" adheres to the weights fairly.  If you have any ideas to further improve the ergonomics, or the complexity, let me know.

AltSysrq commented:
Latest changes look reasonable to me

jentfoo commented:
@AltSysrq This has now been released as 0.14 to maven central

I did update `AuroraClusterMonitor.setReplicaWeight` to be `setHostWeight` after considering that weights for the master will also likely be set.
